File: csharp/src/AnQL.Core/AnQLParser.cs
using AnQL.Core.Grammar;

namespace AnQL.Core;

public class AnQLParser<T> : IAnQLParser<T>
{
    private readonly IAnQLVisitor<T> _visitor;

    public AnQLParser(IAnQLVisitor<T> visitor)
    {
        _visitor = visitor;
    }

    public T Parse(string input)
    {
        if (string.IsNullOrEmpty(input))
            return _visitor.SuccessQueryResult;

        var anqlParser = AnQL.BuildParser(input);

        return _visitor.Visit(anqlParser.query()) ?? _visitor.SuccessQueryResult;
    }
}

File: typescript/core/src/AnQLParser.ts
import { buildGrammarParser } from './AnQL'
import './grammar'
import IAnQLVisitor from './grammar/IAnQLVisitor'

export default class AnQLParser<T> {
    private readonly visitor: IAnQLVisitor<T>

    constructor(visitor: IAnQLVisitor<T>) {
        this.visitor = visitor
    }

    public parse(input: string): T {
        if (input == '') {
            return this.visitor.successQueryResult
        }

        const anqlParser = buildGrammarParser(input)

        return this.visitor.visit(anqlParser.query()) ?? this.visitor.successQueryResult
    }
}